cf-custom-resources/lib/custom-domain-app-runner.js

```diff
@@ -0,0 +1,315 @@
+                await waitForCustomDomainToBeActive(serviceARN, customDomain);
+                break;
+            case "Update":
+            case "Delete":
+                throw new Error("not yet implemented");
+            default:
+                throw new Error(`Unsupported request type ${event.RequestType}`);
+        }
+    };
+
+    try {
+        await Promise.race([exports.deadlineExpired(), handler(),]);
+        await report(event, context, "SUCCESS", physicalResourceID);
+    } catch (err) {
+        if (err.name === ERR_NAME_INVALID_REQUEST && err.message.includes(`${customDomain} is already associated with`)) {
+            await report(event, context, "SUCCESS", physicalResourceID);
+            return;
+        }
+        console.log(`Caught error for service ${serviceARN}: ${err.message}`);
+        await report(event, context, "FAILED", physicalResourceID, null, err.message);
+    }
+};
+
+exports.deadlineExpired = function () {
+    return new Promise(function (resolve, reject) {
+        setTimeout(
+            reject,
+            14 * 60 * 1000 + 30 * 1000 /* 14.5 minutes*/,
+            new Error("Lambda took longer than 14.5 minutes to update environment")
+        );
+    });
+};
+
+/**
+ * Validate certificates of the custom domain for the service by upserting validation records.
+ * Errors are not handled and are directly passed to the caller.
+ *
+ * @param {string} serviceARN ARN of the service that the custom domain applies to.
+ * @param {string} customDomainName the custom domain name.
+ */
+async function addCustomDomain(serviceARN, customDomainName) {
+    const data = await appRunnerClient.associateCustomDomain({
+        DomainName: customDomainName,
+        ServiceArn: serviceARN,
+    }).promise();
+
+    return Promise.all([
+        updateCNAMERecordAndWait(customDomainName, data.DNSTarget, appHostedZoneID, "UPSERT"), // Upsert the record that maps `customDomainName` to the DNS of the app runner service.
+        validateCertForDomain(serviceARN, customDomainName),
+    ]);
+}
+
+/**
+ * Wait for the custom domain to be ACTIVE.
+ * @param {string} serviceARN the service to which the domain is added.
+ * @param {string} customDomainName the domain name.
+ */
+async function waitForCustomDomainToBeActive(serviceARN, customDomainName) {
+    let i;
+    for (i = 0; i < ATTEMPTS_WAIT_FOR_ACTIVE; i++) {
+        const data = await appRunnerClient.describeCustomDomains({
+            ServiceArn: serviceARN,
+        }).promise().catch(err => {
+            throw new Error(`wait for domain ${customDomainName} to be active: ` + err.message);
+        });
+
+        let domain;
+        for (const d of data.CustomDomains) {
+            if (d.DomainName === customDomainName) {
+                domain = d;
+                break;
+            }
+        }
+
+        if (!domain) {
+            throw new Error(`wait for domain ${customDomainName} to be active: : domain ${customDomainName} is not associated`);
```

Comment:
Should it be
```suggestion
            throw new Error(`wait for domain ${customDomainName} to be active: domain ${customDomainName} is not associated`);
```

Reply:
Also just curious do you know in what circumstance we'll throw this error? Like why `customDomainName` is not in `data`?

Reply:
In any normal workflow this error shouldn't be reached, because the call is made almost immediately after `AssociateCustomDomain`, which if returns without error, should be an indication that the domain IS associated.

However I just want to deal with the very very edgy edge case where some other unknown operations disassociate the domain right between the calls...

Reply:
yeah sure i totally agree we should catch this. Just curious...